File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int i=0;
    int ans;
    public int kthSmallest(TreeNode root, int k) {
        ans = 0;
        i=0;
        util(root, k);
        return ans;
    }
    public void util(TreeNode root, int k){
        if(root == null) return ;
        util(root.left, k);
        i++;
        if(i== k){
            ans = root.val;
            return;
        }
        util(root.right,k);

    }
}

File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        int i = 0;
        return util(root, k, i);
    }
    int util(TreeNode* root, int k, int &i){
        if(root == nullptr)
            return -1;
        int left = util(root->left, k, i);
        if(left != -1)
            return left;
        i++;
        if(i == k)
            return root->val;
        return util(root->right, k, i);
        
    }
};